plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/BaseJdbcConnectionCreationTest.java

```diff
@@ -0,0 +1,169 @@
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static io.trino.testing.DataProviders.toDataProvider;
+import static java.util.Collections.synchronizedMap;
+import static java.util.Objects.requireNonNull;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@Test(singleThreaded = true) // this class is stateful, see fields
+public abstract class BaseJdbcConnectionCreationTest
+        extends AbstractTestQueryFramework
+{
+    protected ConnectionCountingConnectionFactory connectionFactory;
+
+    @BeforeClass
+    public void verifySetup()
+    {
+        requireNonNull(connectionFactory, "connectionFactory is null");
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    @Test(dataProvider = "testCases")
+    public void test(TestCase testCase)
+    {
+        int before = connectionFactory.openConnections.get();
+        if (testCase.errorMessage.isPresent()) {
+            assertQueryFails(testCase.query, testCase.errorMessage.get());
+        }
+        else {
+            getQueryRunner().execute(testCase.query);
+        }
+        int after = connectionFactory.openConnections.get();
+        assertThat(after - before).isEqualTo(testCase.expectedJdbcConnectionsCount);
+        connectionFactory.assertThatNoConnectionHasLeaked();
+    }
+
+    @DataProvider
+    public final Object[][] testCases()
+    {
+        return getTestCases().stream()
+                .collect(toDataProvider());
+    }
+
+    protected abstract List<TestCase> getTestCases();
+
+    public static final record TestCase(
+            @Language("SQL") String query,
+            int expectedJdbcConnectionsCount,
+            Optional<String> errorMessage)
+    {
+        public TestCase(@Language("SQL") String query, int expectedJdbcConnectionsCount)
+        {
+            this(query, expectedJdbcConnectionsCount, Optional.empty());
+        }
+
+        public TestCase(@Language("SQL") String query, int expectedJdbcConnectionsCount, String errorMessage)
+        {
+            this(query, expectedJdbcConnectionsCount, Optional.of(errorMessage));
+        }
+
+        public TestCase
+        {
+            requireNonNull(query, "query is null");
+            requireNonNull(errorMessage, "errorMessage is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return query;
+        }
+    }
+
+    protected static class ConnectionCountingConnectionFactory
+            implements ConnectionFactory
+    {
+        // Map from connection to a fake exception indicating where the connection was acquired
+        private final Map<Connection, Exception> connectionCreations = synchronizedMap(new IdentityHashMap<>());
+        private final AtomicInteger openConnections = new AtomicInteger();
```

Comment:
can `connectionCreations.size()` be reused for counting `openConnections ` ?

Reply:
nope, because connections are also being closed.
`connectionCreations` tracks only connections that are currently open

Reply:
`openConnection` -> `openConnectionRequests`
`connectionCreations` -> `connectionCreationPointByOpenConnections`

Reply:
`openConnection` -> `openConnections` is fine too :)